compiler/src/dmd/timetrace.d

```diff
@@ -0,0 +1,462 @@
+Source:    $(LINK2 https://github.com/dlang/dmd/blob/master/src/dmd/common/timetrace.d, common/_timetrace.d)
+Documentation: https://dlang.org/phobos/dmd_common_timetrace.html
+Coverage:    https://codecov.io/gh/dlang/dmd/src/master/src/dmd/common/timetrace.d
+*/
+module dmd.timetrace;
+
+import dmd.location;
+import dmd.dsymbol;
+import dmd.expression;
+import dmd.root.array;
+import dmd.common.outbuffer;
+import dmd.root.string : toDString;
+
+// Thread local profiler instance (multithread currently not supported because compiler is single-threaded)
+TimeTraceProfiler* timeTraceProfiler = null;
+
+/**
+ * Initialize time tracing functionality.
+ *
+ * Must be called before any other calls to timeTrace functions.
+ *
+ * Params:
+ *   timeGranularityUs = minimum event size in microseconds
+ *   processName = name of this executable
+ */
+extern (C++)
+void initializeTimeTrace(uint timeGranularityUs, const(char)* processName)
+{
+    assert(timeTraceProfiler is null, "Double initialization of timeTraceProfiler");
+    timeTraceProfiler = new TimeTraceProfiler(timeGranularityUs, processName);
+}
+
+/**
+ * Cleanup for time tracing functionality.
+ *
+ * After this, no more calls to timeTrace functions can be made.
+ */
+extern (C++)
+void deinitializeTimeTrace()
+{
+    if (timeTraceProfilerEnabled())
+    {
+        object.destroy(timeTraceProfiler);
+        timeTraceProfiler = null;
+    }
+}
+
+/**
+ * Returns: Whether time tracing is enabled.
+ */
+pragma(inline, true)
+extern (C++)
+bool timeTraceProfilerEnabled()
+{
+    version (LDC)
+    {
+        import ldc.intrinsics : llvm_expect;
+
+        return llvm_expect(timeTraceProfiler !is null, false);
+    }
+    else
+    {
+        return timeTraceProfiler !is null;
+    }
+}
+
+/**
+ * Write all time tracing results so far to JSON, in the Chrome Trace Event Format.
+ * Params:
+ *   buf = output buffer to write JSON into
+ */
+extern (C++)
+void writeTimeTraceProfile(OutBuffer* buf)
+{
+    timeTraceProfiler.writeToBuffer(*buf);
+}
+
+/**
+ * Start a new time trace event (C++ interface using upfront C-strings instead of lazy delegates)
+ *
+ * Params:
+ *   name_ptr = event name, visible in high level profile view
+ *   detail_ptr = further details, visible when this event is selected
+ *   loc = source location corresponding to this event
+ */
+extern (C++)
+void timeTraceBeginEvent(scope const(char)* name_ptr, scope const(char)* detail_ptr, Loc loc)
+{
+    import dmd.root.rmem : xarraydup;
+
+    assert(timeTraceProfiler);
+
+    timeTraceProfiler.beginScope(
+        xarraydup(name_ptr.toDString()),
+        xarraydup(detail_ptr.toDString()), loc
+    );
+}
+
+/**
+ * Start a new time trace event
+ *
+ * Details of the event will be passed as delegates to `timeTraceEndEvent` so
+ * they're only generated when the event is actually written.
+ */
+void timeTraceBeginEvent()
+{
+    if (timeTraceProfilerEnabled)
+        timeTraceProfiler.beginScope(null, null, Loc.initial);
```

Comment:
```suggestion
    if (timeTraceProfilerEnabled)
      version (IN_GCC)
        timeTraceBeginEvent(eventType);
      else
        timeTraceProfiler.beginScope(null, null, Loc.initial);
```
Likewise it should be `extern(C++)`.

Saying that if the API is fine, then I could just provide my own implementation of this module that contains none of the private parts.